File: src/test/java/com/back/mozu/domain/reservation/service/DynamicReleaseSchedulerTest.java
package com.back.mozu.domain.reservation.service;

import com.back.mozu.domain.reservation.entity.Reservation;
import com.back.mozu.domain.reservation.entity.ReservationStatus;
import com.back.mozu.domain.reservation.entity.TimeSlot;
import com.back.mozu.domain.reservation.repository.ReservationRepository;
import com.back.mozu.domain.reservation.repository.TimeSlotRepository;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;


import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;

@SpringBootTest
class DynamicReleaseSchedulerTest {

    @Autowired
    private DynamicReleaseScheduler dynamicReleaseScheduler;

    @Autowired
    private ReservationRepository reservationRepository;

    @Autowired
    private TimeSlotRepository timeSlotRepository;

    // 정적 스케줄러 자동실행 막기
    @com.back.mozu.domain.reservation.service.MockBean
    private StaticReleaseScheduler staticReleaseScheduler;

    @AfterEach
    void cleanUp() {
        reservationRepository.deleteAllInBatch();
        timeSlotRepository.deleteAllInBatch();
    }

    @Test
    @DisplayName("동적 스케줄러 - releaseAt 이후 재고 복구")
    void dynamic_scheduler_releases_stock_after_releaseAt() {
        UUID userId = UUID.randomUUID();
        TimeSlot timeSlot = createTimeSlot(8, LocalDate.now().plusMonths(1), LocalTime.of(12, 0));

        Reservation reservation = createReservation(
                userId, timeSlot, 2,
                ReservationStatus.CANCEL_PENDING,
                null, "테스트취소",
                LocalDateTime.now().minusHours(1),
                LocalDateTime.now().minusMinutes(1)
        );

        dynamicReleaseScheduler.releaseStock(reservation.getId());

        Reservation result = reservationRepository.findById(reservation.getId()).orElseThrow();
        TimeSlot updatedSlot = timeSlotRepository.findById(timeSlot.getId()).orElseThrow();

        assertThat(result.getStatus()).isEqualTo(ReservationStatus.CANCELED);
        assertThat(updatedSlot.getStock()).isEqualTo(10); // 8 + 2 = 10
    }

    private TimeSlot createTimeSlot(int stock, LocalDate date, LocalTime time) {
        return timeSlotRepository.save(TimeSlot.builder()
                .date(date).time(time).stock(stock).build());
    }

    private Reservation createReservation(
            UUID userId, TimeSlot timeSlot, int guestCount,
            ReservationStatus status, LocalDateTime cancelledAt,
            String cancelReason, LocalDateTime reservationOpenedAt,
            LocalDateTime releaseAt) {
        return reservationRepository.save(Reservation.builder()
                .userId(userId).timeSlot(timeSlot).guestCount(guestCount)
                .status(status).cancelledAt(cancelledAt).cancelReason(cancelReason)
                .reservationOpenedAt(reservationOpenedAt).releaseAt(releaseAt).build());
    }
}

File: src/test/java/com/back/mozu/domain/reservation/service/DynamicReleaseSchedulerTest.kt
package com.back.mozu.domain.reservation.service

import com.back.mozu.domain.reservation.entity.Reservation
import com.back.mozu.domain.reservation.entity.ReservationStatus
import com.back.mozu.domain.reservation.entity.TimeSlot
import com.back.mozu.domain.reservation.repository.ReservationRepository
import com.back.mozu.domain.reservation.repository.TimeSlotRepository
import io.mockk.Runs
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.just
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.scheduling.TaskScheduler
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneId
import java.util.UUID
import java.util.concurrent.ScheduledFuture

@ExtendWith(MockKExtension::class)
class DynamicReleaseSchedulerTest {

    @MockK
    lateinit var taskScheduler: TaskScheduler

    @MockK
    lateinit var reservationRepository: ReservationRepository

    @MockK
    lateinit var timeSlotRepository: TimeSlotRepository

    @MockK
    private lateinit var releaseStockService: ReleaseStockService

    @InjectMockKs
    lateinit var dynamicReleaseScheduler: DynamicReleaseScheduler

    @Test
    fun `동적 스케줄러는 releaseAt 이후 재고를 복구해야 한다`() {
        // given
        val reservationId = UUID.randomUUID()
        val timeSlotId = UUID.randomUUID()

        val timeSlot = TimeSlot(
            date = LocalDate.now().plusMonths(1),
            time = LocalTime.of(12, 0),
            stock = 8,
            id = timeSlotId,
        )

        val reservation = Reservation(
            id = reservationId,
            userId = UUID.randomUUID(),
            timeSlot = timeSlot,
            guestCount = 2,
            status = ReservationStatus.CANCEL_PENDING,
            cancelledAt = null,
            cancelReason = "테스트취소",
            reservationOpenedAt = LocalDateTime.now().minusHours(1),
            releaseAt = LocalDateTime.now().minusMinutes(1),
        )

        every {
            reservationRepository.findAllByStatus(ReservationStatus.CANCEL_PENDING)
        } returns listOf(reservation)
        every {
            taskScheduler.schedule(any(), any<Instant>())
        } returns mockk()
        every {
            releaseStockService.releaseStock(any())
        } just Runs

        // when
        dynamicReleaseScheduler.run(mockk())

        // then
        verify(exactly = 1) { taskScheduler.schedule(any(), any<Instant>()) }
    }

    @Test
    fun `releaseAt은 Asia Seoul 시간 기준으로 스케줄링되어야 한다`() {
        // given
        val reservationId = UUID.randomUUID()
        val releaseAt = LocalDateTime.of(2026, 5, 18, 12, 0)
        var scheduledInstant: Instant? = null

        val reservation = Reservation(
            id = reservationId,
            status = ReservationStatus.CANCEL_PENDING,
            releaseAt = releaseAt,
        )

        val scheduledFuture = mockk<ScheduledFuture<*>>(relaxed = true)

        every {
            taskScheduler.schedule(any(), any<Instant>())
        } answers {
            scheduledInstant = secondArg()
            scheduledFuture
        }

        // when
        dynamicReleaseScheduler.schedule(reservation)

        // then
        assertThat(scheduledInstant)
            .isEqualTo(releaseAt.atZone(ZoneId.of("Asia/Seoul")).toInstant())
    }
}